File: 2021/day1/scala/solution.scala
import scala.io.Source

@main
def main() = {
  println("Part 1: ")
  print("Example: ")
  part1("../example.txt")
  print("Input: ")
  part1("../input.txt")
  println()

  println("Part 2: ")
  print("Example: ")
  part2("../example.txt")
  print("Input: ")
  part2("../input.txt")
  println()
}

def part1(path : String) = {
  
  val input = Source.fromFile(path)
                  .getLines
                  .map(line => line.toInt)
                  .toList

  // val result = countIncreasing(input)
  val result = input.sliding(2).count { 
    case Seq(a, b) => a < b
    case _ => false
  }
  println(result)
}

def part2(path: String) = {
  val input = Source.fromFile(path)
                  .getLines
                  .map(line => line.toInt)
                  .toList

  val result = input.sliding(3).map(_.sum).sliding(2).count {
    case Seq(a, b) => a < b
    case _ => false
  }
  
  println(result)
}

File: 2021/day1/ruby/solution.rb
input = File.readlines(__dir__ + '/../input.txt')
example = File.readlines(__dir__ + '/../example.txt')

def part1(input)
  input.unshift(input.first)
       .map(&:to_i)
       .each_cons(2)
       .filter { _1 < _2 }
       .count
end

def part2(input)
  summed = input.map(&:to_i)
       .each_cons(3)
       .map(&:sum)

  part1(summed) - 1
end

puts "Part 1: #{part1(input)}"
puts "Part 2: #{part2(input)}"
